lib/Sema/TypeCheckType.cpp

```diff
@@ -1061,6 +1061,8 @@ static Type diagnoseUnknownType(TypeResolution resolution,
         // type. Fix this by replacing 'Self' with the nominal type name.
         assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
 
+        return nominal->getDeclaredInterfaceType();
```

Comment:
I would handle this in the caller of diagnoseUnknownType(), since you're not diagnosing here -- this is valid code now.

Reply:
There are a couple of places diagnoseUnknownType() is called.
```C++
static Type
resolveTopLevelIdentTypeComponent(TypeResolution resolution,
                                  ComponentIdentTypeRepr *comp,
                                  TypeResolutionOptions options) {
```
and
```C++
static Type resolveNestedIdentTypeComponent(
              TypeResolution resolution,
              Type parentTy,
              SourceRange parentRange,
              ComponentIdentTypeRepr *comp,
              TypeResolutionOptions options) {
```
I'm not sure which is the critical path.

Reply:
That's right, only the former path is relevant. "Self" cannot appear as a member type of another type.